### PR TITLE
[tests-only][full-ci] adding test for deleting auth-app token using the token received from auth-app list request

### PR DESCRIPTION
**Pull request by ghost**
## Description

This PR adds test for deleting auth-app token using the token received from auth-app list request 

## Related Issue

- part of: https://github.com/owncloud/ocis/issues/10619

## How Has This Been Tested?

- test environment:
- test case 1: locally
- test case 2: ci
- ...

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

File: tests/acceptance/TestHelpers/AuthAppHelper.php [hunk 1]
```diff
@@ -105,4 +105,29 @@ public static function deleteAppAuthToken(
...
+	 * @return ResponseInterface
+	 */
+	public static function createAuthAppTokenViaImpersonationAPI(
+		string $baseUrl,
+		string $user,
+		string $password,
+		string $expiration,
+		string $impersonatorUser
+	): ResponseInterface {
+		$url = $baseUrl . self::getAuthAppEndpoint() . "?expiry=$expiration&userName=$impersonatorUser";
```
Comment (saw-jan) [inline tests/acceptance/TestHelpers/AuthAppHelper.php:125]: this should be fixed is one of your PR

**Review by saw-jan — CHANGES_REQUESTED** [review]

File: tests/acceptance/bootstrap/AuthAppContext.php [hunk 2]
```diff
@@ -154,4 +156,45 @@ public function theAdministratorCreatesAppTokenForUserWithExpirationTimeViaAuthA
 			)
 		);
 	}
+
+	/**
+	 * @When user :user delets all created auth-app token
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/AuthAppContext.php:161]:
```suggestion
	 * @When user :user deletes all created auth-app token
```

File: tests/acceptance/bootstrap/AuthAppContext.php [hunk 3]
```diff
@@ -154,4 +156,45 @@ public function theAdministratorCreatesAppTokenForUserWithExpirationTimeViaAuthA
...
+	 *
+	 * @param string $user
+	 *
+	 * @return void
+	 */
+	public function userDeletsAllCreatedAuthAppToken(string $user): void {
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/AuthAppContext.php:167]:
```suggestion
	public function userDeletesAllCreatedAuthAppToken(string $user): void {
```

File: tests/acceptance/bootstrap/AuthAppContext.php [hunk 4]
```diff
@@ -154,4 +156,45 @@ public function theAdministratorCreatesAppTokenForUserWithExpirationTimeViaAuthA
...
+				$this->featureContext->getActualUsername($user),
+				$this->featureContext->getPasswordForUser($user),
+				$response->token
+			);
+			$this->lastHttpStatusCodesArray[] = $deleteResponse->getStatusCode();
+		}
+	}
+
+	/**
+	 * @Then the HTTP status code of responses on each endpoint should be :statusCodes
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/AuthAppContext.php:186]: there should be a similar existing step.

File: tests/acceptance/features/apiAuthApp/token.feature [hunk 5]
```diff
@@ -129,3 +129,12 @@ Feature: create auth-app token
         }
       }
       """
+
+  @env-config
+  Scenario: user deleats auth-app token
```
Comment (prashant-gurung899) [inline tests/acceptance/features/apiAuthApp/token.feature:134]:
```suggestion
  Scenario: user deletes auth-app token
```

**Review by amrita-shrestha — CHANGES_REQUESTED** [review]
why is your PR missing information about https://github.com/owncloud/ocis/issues/10675. So confusing. I think you need to update bug report title to be more specific and this PR title too.

File: tests/acceptance/features/apiAuthApp/token.feature [hunk 6]
```diff
@@ -129,3 +129,12 @@ Feature: create auth-app token
         }
       }
       """
+
+  @env-config
+  Scenario: user deletes auth-app token
```
Comment (amrita-shrestha) [inline tests/acceptance/features/apiAuthApp/token.feature:134]: > Can't delete auth-app token using the token received from create request but deleting with the token that we get when listing works.

 It seems multiple ways are present to delete auth-app token. Please specify that in scenario

File: tests/acceptance/bootstrap/AuthAppContext.php [hunk 7]
```diff
@@ -154,4 +155,29 @@ public function theAdministratorCreatesAppTokenForUserWithExpirationTimeViaAuthA
...
+
+	/**
+	 * @When user :user deletes all created auth-app token
+	 *
+	 * @param string $user
+	 *
+	 * @return void
+	 */
+	public function userdeletesAllCreatedAuthAppToken(string $user): void {
+		$responses = AuthAppHelper::listAllAppAuthTokensForUser(
```
Comment (amrita-shrestha) [inline tests/acceptance/bootstrap/AuthAppContext.php:167]:
```suggestion
		$authAppTokens = AuthAppHelper::listAllAppAuthTokensForUser(
```

File: tests/acceptance/bootstrap/AuthAppContext.php [hunk 8]
```diff
@@ -154,4 +155,29 @@ public function theAdministratorCreatesAppTokenForUserWithExpirationTimeViaAuthA
...
+		);
+		$responses = json_decode($responses->getBody()->getContents());
+		foreach ($responses as $response) {
+			$deleteResponse = AuthAppHelper::deleteAppAuthToken(
+				$this->featureContext->getBaseUrl(),
+				$this->featureContext->getActualUsername($user),
+				$this->featureContext->getPasswordForUser($user),
+				$response->token
+			);
+			$this->featureContext->pushToLastHttpStatusCodesArray((string)$deleteResponse->getStatusCode());
```
Comment (amrita-shrestha) [inline tests/acceptance/bootstrap/AuthAppContext.php:180]: is status code datatype **int**?

File: tests/acceptance/bootstrap/AuthAppContext.php [hunk 9]
```diff
@@ -154,4 +155,29 @@ public function theAdministratorCreatesAppTokenForUserWithExpirationTimeViaAuthA
...
+	 *
+	 * @return void
+	 */
+	public function userdeletesAllCreatedAuthAppToken(string $user): void {
+		$responses = AuthAppHelper::listAllAppAuthTokensForUser(
+			$this->featureContext->getBaseUrl(),
+			$this->featureContext->getActualUsername($user),
+			$this->featureContext->getPasswordForUser($user),
+		);
+		$responses = json_decode($responses->getBody()->getContents());
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/AuthAppContext.php:172]:
```suggestion
		$response = AuthAppHelper::listAllAppAuthTokensForUser(
			$this->featureContext->getBaseUrl(),
			$this->featureContext->getActualUsername($user),
			$this->featureContext->getPasswordForUser($user),
		);
		$authAppTokens = json_decode($responses->getBody()->getContents());
```

File: tests/acceptance/bootstrap/AuthAppContext.php [hunk 10]
```diff
@@ -154,4 +155,29 @@ public function theAdministratorCreatesAppTokenForUserWithExpirationTimeViaAuthA
...
+	 * @return void
+	 */
+	public function userDeletesAllCreatedAuthAppToken(string $user): void {
+		$response = AuthAppHelper::listAllAppAuthTokensForUser(
+			$this->featureContext->getBaseUrl(),
+			$this->featureContext->getActualUsername($user),
+			$this->featureContext->getPasswordForUser($user),
+		);
+		$authAppTokens = json_decode($response->getBody()->getContents());
+		foreach ($authAppTokens as $response) {
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/AuthAppContext.php:173]:
```suggestion
		foreach ($authAppTokens as $tokenObj) {
```

File: tests/acceptance/features/apiAuthApp/token.feature [hunk 11]
```diff
@@ -129,3 +129,12 @@ Feature: create auth-app token
...
       }
       """
+
+  @env-config
+  Scenario: user deletes auth-app token using token form list request
+    Given the config "AUTH_APP_ENABLE_IMPERSONATION" has been set to "true"
+    And user "Alice" has created app token with expiration time "72h" using the auth-app API
+    And the administrator has created app token for user "Alice" with expiration time "72h" using the auth-app API
+    And user "Alice" has created app token with expiration time "72h" using the auth-app CLI
+    When user "Alice" deletes all created auth-app token
```
Comment (amrita-shrestha) [inline tests/acceptance/features/apiAuthApp/token.feature:139]: for future proof, i think we should specify which method used to delete

Comment (saw-jan) [reply tests/acceptance/features/apiAuthApp/token.feature]: there's no different way to delete app tokens.

File: tests/acceptance/features/apiAuthApp/token.feature [hunk 12]
```diff
@@ -129,3 +129,12 @@ Feature: create auth-app token
...
       }
       """
+
+  @env-config
+  Scenario: user deletes auth-app token using token form list request
+    Given the config "AUTH_APP_ENABLE_IMPERSONATION" has been set to "true"
+    And user "Alice" has created app token with expiration time "72h" using the auth-app API
+    And the administrator has created app token for user "Alice" with expiration time "72h" using the auth-app API
+    And user "Alice" has created app token with expiration time "72h" using the auth-app CLI
+    When user "Alice" deletes all created auth-app token using the token token received from auth-app list request
```
Comment (saw-jan) [inline tests/acceptance/features/apiAuthApp/token.feature:139]:
```suggestion
    When user "Alice" deletes all the created auth-app tokens using the auth-app API
```

**Review by saw-jan — CHANGES_REQUESTED** [review]

File: tests/acceptance/bootstrap/AuthAppContext.php [hunk 13]
```diff
@@ -154,4 +155,29 @@ public function theAdministratorCreatesAppTokenForUserWithExpirationTimeViaAuthA
...
 	}
+
+	/**
+	 * @When user :user deletes all the created auth-app tokens using the auth-app API
+	 *
+	 * @param string $user
+	 *
+	 * @return void
+	 */
+	public function userDeletesAllCreatedAuthAppToken(string $user): void {
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/AuthAppContext.php:166]:
```suggestion
	public function userDeletesAllCreatedAuthAppTokenUsingAuthAppAPI(string $user): void {
```

File: tests/acceptance/features/apiAuthApp/token.feature (hunk 11 above)
Comment (amrita-shrestha) [reply tests/acceptance/features/apiAuthApp/token.feature]: I think you should be more clear in your comment, @saw-jan. you mean this bug report is actual behavior https://github.com/owncloud/ocis/issues/10675 so there's only one way to delete app token

**Review by amrita-shrestha — CHANGES_REQUESTED** [review]
IMO its better if other edge case added too in this PR

File: tests/acceptance/bootstrap/AuthAppContext.php [hunk 14]
```diff
@@ -154,4 +155,29 @@ public function theAdministratorCreatesAppTokenForUserWithExpirationTimeViaAuthA
...
+	 *
+	 * @param string $user
+	 *
+	 * @return void
+	 */
+	public function userDeletesAllCreatedAuthAppTokenUsingAuthAppAPI(string $user): void {
+		$response = AuthAppHelper::listAllAppAuthTokensForUser(
+			$this->featureContext->getBaseUrl(),
+			$this->featureContext->getActualUsername($user),
+			$this->featureContext->getPasswordForUser($user),
```
Comment (nabim777) [inline tests/acceptance/bootstrap/AuthAppContext.php:170]:
```suggestion
			$this->featureContext->getBaseUrl(),
			$this->featureContext->getActualUsername($user),
			$this->featureContext->getPasswordForUser($user),
```
these are repeated function calls
IMO, it will good to store in local variable and use it.

**Comment by nabim777** [comment]
Is it possible to add a `then` step that ensures the token does not exist or is deleted?

File: tests/acceptance/features/apiAuthApp/token.feature [hunk 15]
```diff
@@ -129,3 +129,12 @@ Feature: create auth-app token
         }
       }
       """
+
+  @env-config
+  Scenario: user deletes auth-app token using token form list request
```
Comment (saw-jan) [inline tests/acceptance/features/apiAuthApp/token.feature:134]:
```suggestion
  Scenario: user deletes the created auth-app token
```

File: tests/acceptance/features/apiAuthApp/token.feature [hunk 16]
```diff
@@ -129,3 +129,12 @@ Feature: create auth-app token
...
+    Given the config "AUTH_APP_ENABLE_IMPERSONATION" has been set to "true"
+    And user "Alice" has created app token with expiration time "72h" using the auth-app API
+    And the administrator has created app token for user "Alice" with expiration time "72h" using the auth-app API
+    And user "Alice" has created app token with expiration time "72h" using the auth-app CLI
```
Comment (saw-jan) [inline tests/acceptance/features/apiAuthApp/token.feature:138]: let's not use CLI step in API tests. create separate scenario for CLI in cli test suite
```suggestion
    And the administrator has created app token for user "Alice" with expiration time "72h" using the auth-app API
```

File: tests/acceptance/features/apiAuthApp/token.feature [hunk 17]
```diff
@@ -129,3 +111,12 @@ Feature: create auth-app token
...
       """
+
+  @env-config
+  Scenario: user deletes the created auth-app token
+    Given the config "AUTH_APP_ENABLE_IMPERSONATION" has been set to "true"
+    And user "Alice" has created app token with expiration time "72h" using the auth-app API
+    And the administrator has created app token for user "Alice" with expiration time "72h" using the auth-app API
+    When user "Alice" deletes all the created auth-app tokens using the auth-app API
+    Then the HTTP status code of responses on each endpoint should be "200, 200" respectively
+    And user "Alice" should have "0" auth-app tokens
```
Comment (amrita-shrestha) [inline tests/acceptance/features/apiAuthApp/token.feature:122]: i think json response check also good

**Review by saw-jan — DISMISSED** [review]

File: tests/acceptance/bootstrap/AuthAppContext.php [hunk 18]
```diff
@@ -154,4 +155,65 @@ public function theAdministratorCreatesAppTokenForUserWithExpirationTimeViaAuthA
...
+
+	/**
+	 * @Then user :user should have :count auth-app tokens
+	 *
+	 * @param string $user
+	 * @param integer $count
+	 *
+	 * @return void
+	 */
+	public function thereShouldBeAuthAppTokens(string $user, int $count): void {
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/AuthAppContext.php:200]:
```suggestion
	public function userShouldHaveAuthAppTokens(string $user, int $count): void {
```

File: tests/acceptance/bootstrap/AuthAppContext.php [hunk 19]
```diff
@@ -154,4 +155,65 @@ public function theAdministratorCreatesAppTokenForUserWithExpirationTimeViaAuthA
...
+			$this->featureContext->getBaseUrl(),
+			$this->featureContext->getActualUsername($user),
+			$this->featureContext->getPasswordForUser($user),
+		);
+		$authAppTokens = $response->getBody()->getContents();
+		if ($authAppTokens === '[]') {
+			$authAppTokens = [];
+		} else {
+			$authAppTokens = json_decode($response->getBody()->getContents());
+		}
```
Comment (saw-jan) [inline tests/acceptance/bootstrap/AuthAppContext.php:211]:
```suggestion
		$authAppTokens = json_decode($response->getBody()->getContents());
```

**Review by saw-jan — APPROVED** [review]

**Review by amrita-shrestha — CHANGES_REQUESTED** [review]
Did you checked edge cases like 
- admin creating app-token for alice user and alice user deleting that toke
- alice user creating token for him/her self and admin deleting that token
- admin creating token for alice, brian tries to delete alice token

**Comment by saw-jan** [comment]
> Did you checked edge cases like
> 
> * admin creating app-token for alice user and alice user deleting that toke
> * alice user creating token for him/her self and admin deleting that token
> * admin creating token for alice, brian tries to delete alice token

Please, add them to the QA ticket

**Review by amrita-shrestha — APPROVED** [review]
